core-js/src/definitions.ts

```diff
@@ -0,0 +1,1421 @@
+export type InputTypeReferencer = InputFieldDefinition | ArgumentDefinition<any>;
+export type ObjectTypeReferencer = OutputTypeReferencer | UnionType | SchemaDefinition;
+export type InterfaceTypeReferencer = OutputTypeReferencer | ObjectType | InterfaceType;
+
+export type NullableType = NamedType | ListType<any>;
+
+export type NamedTypeKind = NamedType['kind'];
+
+export function defaultRootTypeName(root: SchemaRoot) {
+  return root.charAt(0).toUpperCase() + root.slice(1);
+}
+
+export function isNamedType(type: Type): type is NamedType {
+  return type instanceof BaseNamedType;
+}
+
+export function isWrapperType(type: Type): type is WrapperType {
+  return type.kind == 'ListType' || type.kind == 'NonNullType';
+}
+
+export function isOutputType(type: Type): type is OutputType {
+  if (isWrapperType(type)) {
+    return isOutputType(type.baseType());
+  }
+  switch (type.kind) {
+    case 'ScalarType':
+    case 'ObjectType':
+    case 'UnionType':
+    case 'EnumType':
+    case 'InterfaceType':
+      return true;
+    default:
+      return false;
+  }
+}
+
+export function ensureOutputType(type: Type): OutputType {
+  if (isOutputType(type)) {
+    return type;
+  } else {
+    throw new Error(`Type ${type} (${type.kind}) is not an output type`);
+  }
+}
+
+export function isInputType(type: Type): type is InputType {
+  if (isWrapperType(type)) {
+    return isInputType(type.baseType());
+  }
+  switch (type.kind) {
+    case 'ScalarType':
+    case 'EnumType':
+    case 'InputObjectType':
+      return true;
+    default:
+      return false;
+  }
+}
+
+export function ensureInputType(type: Type): InputType {
+  if (isInputType(type)) {
+    return type;
+  } else {
+    throw new Error(`Type ${type} (${type.kind}) is not an input type`);
+  }
+}
+
+export interface Named {
+  readonly name: string;
+}
+
+export abstract class SchemaElement<Parent extends SchemaElement<any, any> | Schema, Referencer> {
+  protected _parent?: Parent;
+  protected readonly _appliedDirectives: Directive[] = [];
+  description?: string;
+  source?: ASTNode;
+
+  abstract coordinate: string;
+
+  schema(): Schema | undefined {
+    if (!this._parent) {
+      return undefined;
+    } else if ( this._parent instanceof Schema) {
+      return this._parent as any;
+    } else {
+      return (this._parent as SchemaElement<any, any>).schema();
+    }
+  }
+
+  get parent(): Parent | undefined {
+    return this._parent;
+  }
+
+  protected setParent(parent: Parent) {
+    assert(!this._parent, "Cannot set parent of a non-detached element");
+    this._parent = parent;
+  }
+
+  get appliedDirectives(): readonly Directive[] {
+    return this._appliedDirectives;
+  }
+
+  appliedDirective(definition: DirectiveDefinition): Directive[] {
+    return this._appliedDirectives.filter(d => d.name == definition.name);
+  }
+
+  applyDirective(directive: Directive): Directive; 
+  applyDirective(definition: DirectiveDefinition, args?: Map<string, any>): Directive; 
```

Comment:
Although `Map` is preferable for storage nowadays, I think an object (`Record<string, any>`) makes for a more natural API for `args` because you can pass in an object literal.

Reply:
I also think we'll want `DirectiveDefinition<TArgs>` so we can have (optional) strong typing for directive arguments.

Reply:
> I think an object (`Record<string, any>`) makes for a more natural API for args

Fair point.

> I also think we'll want `DirectiveDefinition<TArgs>` so we can have (optional) strong typing for directive arguments.

I'll check how that could work in practice (I can see how strong typing for known directives would be nice; I just wonder how carrying this between `DirectiveDefinition` and `Directive` will work in practice).

Reply:
As I though, I can't really find a way to handle `DirectiveDefinition<TArgs>` in a safe way. I mean, the args of a _definition_ are complex objects that can have directives and whatnot, and so I don't see a way to have `DirectiveDefinition` take a generic that reflect those properly and could be transformed to type-check the applications of that directives.

What I did do is add a `DirectiveDefinition<TApplicationArgs>`, where `TApplicationArgs` is directly the type of _application_ arguments. But this does mean there is no way to ensure (statically or not really) that this generic argument is valid. However, _assuming_ it is valid, it does provide nice type-checking for the related directives and I think that's still sufficiently useful to be worth the slight risk of getting this wrong.

Please have a look at https://github.com/apollographql/federation/pull/823/commits/a969ad569478df0ec79024adc64727a312daa6b0 for what I did do, see if this look anywhere near what you had in mind.